### PR TITLE
3 d tile untile watershed

### PR DESCRIPTION
**Pull request by cdpavelchek**
3D tile/untile and 3D watershed

### CODE REVIEW

File: deepcell_toolbox/deep_watershed.py [hunk 1]
```diff
@@ -283,3 +283,67 @@ def format_output_multiplex(output_list):
...
+        distance_threshold (float): Threshold for the outer distance.
+        exclude_border (bool): Whether to include centroid detections
+            at the border.
+        small_objects_threshold (int): Removes objects smaller than this size.
+
+    Returns:
+        numpy.array: Uniquely labeled mask.
+    """
+    inner_distance_batch = outputs[0][:, ..., 0]
+    outer_distance_batch = outputs[1][:, ..., 0]
```
Comment (willgraf) [inline deepcell_toolbox/deep_watershed.py:317]:
```suggestion
    inner_distance_batch = outputs[0][..., 0]
    outer_distance_batch = outputs[1][..., 0]
```

File: deepcell_toolbox/utils.py [hunk 2]
```diff
@@ -450,3 +450,265 @@ def untile_image(tiles, tiles_info, power=2, **kwargs):
...
+    overlaps_x = []
+    overlaps_y = []
+
+    for b in range(image.shape[0]):
+        for i in range(rep_number_x):
+            for j in range(rep_number_y):
+                for k in range(rep_number_z):
+                    z_axis = 1
+                    x_axis = 2
+                    y_axis = 3
```
Comment (willgraf) [inline deepcell_toolbox/utils.py:531]: Define these axes outside of the loop.

File: deepcell_toolbox/utils.py [hunk 3]
```diff
@@ -450,3 +450,265 @@ def untile_image(tiles, tiles_info, power=2, **kwargs):
...
+    return window
+
+
+def untile_image_3D(tiles, tiles_info, power=3, force=False, **kwargs):
+    """Untile a set of tiled images back to the original model shape.
+
+     Args:
+         tiles (numpy.array): The tiled images image to untile.
+         tiles_info (dict): Details of how the image was tiled (from tile_image).
+         power (int): The power of the window function
```
Comment (willgraf) [inline deepcell_toolbox/utils.py:645]: add `force` to the docstring. Why would we set it to `True`?

File: deepcell_toolbox/utils.py [hunk 4]
```diff
@@ -450,3 +450,265 @@ def untile_image(tiles, tiles_info, power=2, **kwargs):
...
+    overlaps_z = tiles_info['overlaps_z']
+
+    tile_size_x = tiles_info['tile_size_x']
+    tile_size_y = tiles_info['tile_size_y']
+    tile_size_z = tiles_info['tile_size_z']
+    pad_x = tiles_info['pad_x']
+    pad_y = tiles_info['pad_y']
+    pad_z = tiles_info['pad_z']
+
+    image_shape = [image_shape[0], image_shape[1], image_shape[2], image_shape[3], tiles.shape[-1]]
```
Comment (willgraf) [inline deepcell_toolbox/utils.py:681]:
```suggestion
    image_shape = tuple(list(image_shape[:4]) + [tiles.shape[-1]])
```

File: deepcell_toolbox/utils_test.py [hunk 5]
```diff
@@ -302,3 +302,112 @@ def test_untile_image():
...
+        if stride_y > tile_size_y:
+            stride_y = tile_size_y
+
+        rep_number_x = ceil((image_size_x - tile_size_x) / stride_x + 1)
+        rep_number_y = ceil((image_size_y - tile_size_y) / stride_y + 1)
+        rep_number_z = ceil((image_size_z - tile_size_z) / stride_z + 1)
+
+        expected_batches = big_image.shape[0] * rep_number_x * rep_number_y * rep_number_z
+
+        assert tiles.shape[0] == expected_batches
```
Comment (willgraf) [inline deepcell_toolbox/utils_test.py:357]: Is there any way to test this without duplicating the code?

File: deepcell_toolbox/utils.py [hunk 6]
```diff
@@ -450,3 +450,265 @@ def untile_image(tiles, tiles_info, power=2, **kwargs):
...
+    window_size = (tile_size_z, tile_size_x, tile_size_y)
+    image = np.zeros(image_shape, dtype=np.float)
+
+    for (tile, batch, x_start, x_end, y_start, y_end, z_start,
+         z_end, overlap_x, overlap_y, overlap_z) in zip(
+            tiles, batches, x_starts, x_ends, y_starts, y_ends,
+            z_starts, z_ends, overlaps_x, overlaps_y, overlaps_z):
```
Comment (willgraf) [inline deepcell_toolbox/utils.py:688]: Save the output of `zip()` as another variable. This is a little tough to read.

File: deepcell_toolbox/utils_test.py [hunk 7]
```diff
@@ -302,3 +302,112 @@ def test_untile_image():
...
+        utils.tile_image(bad_image, (5, 5), stride_ratio=0.75)
+
+
+def test_untile_image_3D():
+    shapes = [
+        (1, 30, 60, 51, 2),
+        (2, 20, 90, 30, 1)
+    ]
+
+    rand_rel_diff_thresh = 5e-1
```
Comment (willgraf) [inline deepcell_toolbox/utils_test.py:372]: 50%?

File: deepcell_toolbox/utils.py (hunk 2 above)
Comment (cdpavelchek) [reply deepcell_toolbox/utils.py:531]: done

File: deepcell_toolbox/utils.py (hunk 3 above)
Comment (cdpavelchek) [reply deepcell_toolbox/utils.py:645]: Added. And it's so we're able to still use spline interpolation outside the bounds of what is normally defined as acceptable in terms of stride_ratio/tile_size. Should normally be false but I wanted to have that option for some of the 3D stuff.

File: deepcell_toolbox/utils.py (hunk 4 above)
Comment (cdpavelchek) [reply deepcell_toolbox/utils.py:681]: Done - added manually

File: deepcell_toolbox/utils_test.py (hunk 7 above)
Comment (cdpavelchek) [reply deepcell_toolbox/utils_test.py:372]: my bad, supposed to be 2%. Fixed.

File: deepcell_toolbox/utils.py (hunk 6 above)
Comment (cdpavelchek) [reply deepcell_toolbox/utils.py:688]: done

File: deepcell_toolbox/utils.py [hunk 8]
```diff
@@ -450,3 +450,267 @@ def untile_image(tiles, tiles_info, power=2, **kwargs):
...
+     Returns:
+         numpy.array: The untiled image.
+     """
+    # Define mininally acceptable tile_size and stride_ratios for spline interpolation
+    min_tile_size = 32
+    min_stride_ratio = 0.5
+
+    if force:
+        min_tile_size = 0
+        min_tile_height = 0
```
Comment (willgraf) [inline deepcell_toolbox/utils.py:657]:
```suggestion
```

I don't think `min_tile_height` is used anywhere.

File: deepcell_toolbox/utils.py [hunk 9]
```diff
@@ -450,3 +450,267 @@ def untile_image(tiles, tiles_info, power=2, **kwargs):
...
+
+    image_shape = tuple(list(image_shape[:4]) + [tiles.shape[-1]])
+    window_size = (tile_size_z, tile_size_x, tile_size_y)
+    image = np.zeros(image_shape, dtype=np.float)
+
+    tile_data_zip = zip(tiles, batches, x_starts, x_ends, y_starts,
+                        y_ends, z_starts, z_ends, overlaps_x, overlaps_y, overlaps_z)
+
+    for (tile, batch, x_start, x_end, y_start, y_end, z_start,
+         z_end, overlap_x, overlap_y, overlap_z) in tile_data_zip::
```
Comment (willgraf) [inline deepcell_toolbox/utils.py:690]:
```suggestion
         z_end, overlap_x, overlap_y, overlap_z) in tile_data_zip:
```

File: deepcell_toolbox/utils.py (hunk 8 above)
Comment (cdpavelchek) [reply deepcell_toolbox/utils.py:657]: it's not,  meant to take that out

**Pull request merged**